File: code/extensiveHashTable.cpp
/**********************************************
  > File Name		: extensiveHashTable.cpp
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Thu 04 Feb 2021 09:32:37 PM CST
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

#include <stdio.h>
#include <stdlib.h>
#include <cstring>
#include <unistd.h>

#define RED "\e[0;31m"
#define CYAN "\e[0;36m"
#define MAGENTA "\e[0;35m"
#define NONE "\e[0m"

struct Record {
    u_int32_t key;
    char* str;
    Record(u_int32_t key, const char* s) {
        if (strlen(s) > 100) {
            printf("s length out of limits.\n");
            this->str = NULL;
        } else {
            str = new char[10];
            this->key = key;
            strcpy(this->str, s);
        }
    }
};
/* In our case, 4 record pointers per block.*/
struct Block {
    u_int32_t j;
    Record* records[4];
    Block() {
        j = 1;
        for (int k = 0; k < 4; k++) {
            records[k] = NULL;
        }
    }
};


class ExtensiveHashTable {
private:
    u_int32_t i;

    Block** bucketArray;

    char* toBinary(u_int32_t key, u_int32_t length, bool output, bool toPrint);

    char* toReverseBinary(u_int32_t key, u_int32_t length);

    int toDecimal(const char* hash, u_int32_t length);

    char* hash(u_int32_t key);

    void expand();

    Block* findBlock(const u_int32_t key);

    int findLocation(const u_int32_t key);

    void adjustNewBlock(Block* oldBlock, Block* newBlock, int location);

public:
    ExtensiveHashTable(u_int32_t size);
    
    Record* search(const u_int32_t key, Block* b);
    
    void insertRecord(Record* r);

    void deleteRecord(const u_int32_t key);

    void displayTable();
};

/* Transform a integer into its binary form. */
char* ExtensiveHashTable::toBinary(u_int32_t key, u_int32_t length = 32, bool output = true, bool toPrint = false) {
    if (length == 0) {
        printf("toBinary length = 0 is not allowed.\n");
        return NULL;
    }
    u_int32_t temp = key;
    char* res;
    if (!toPrint)
        res = new char[length];
    else {
        res = new char[length+1];
        res[length] = '\0';
    }
    int index = 0;
    while (key && index < length) {
        if (key & 0x00000001) {
            res[index++] = '1';
        } else {
            res[index++] = '0';
        }
        key >>= 1;
    }
    while (index < length) {
        res[index++] = '0';
    }
    if (output) 
        printf("%d to binary: %s\n", temp, res);
    return res;
}

char* ExtensiveHashTable::toReverseBinary(u_int32_t key, u_int32_t length = 32) {
    char* res = new char[length+1];
    res[length] = '\0';
    int index = length-1;
    while (key && index > -1) {
        if (key & 0x00000001) {
            res[index--] = '1';
        } else {
            res[index--] = '0';
        }
        key >>= 1;
    }
    while (index > -1) {
        res[index--] = '0';
    }
    return res;
}

int ExtensiveHashTable::toDecimal(const char* hash, u_int32_t length = 32) {
    if (length == 0) {
        return 0;
    }
    int res = 0;
    for (int k = 0; k < length; k++) {
        if (hash[k] == '0') {
            continue;
        } else if (hash[k] == '1') {
            res += (1<<(length-k-1));
        } else {
            printf(RED "Unknown char in the binary form.\n" NONE);
            return -1;
        }
    }
    return res;
}

/* For simlicity, in our case, I won't implement hash functions in 
 * the cpp standard library.
 * So I create a simple "hash" function with no hashing, I'll just
 * transform a decimal integer into its binary form. And disorder it 
 * in a certain way.
 */
char* ExtensiveHashTable::hash(u_int32_t key) {
    printf("Hash key = %d.\n", key);
    if (key < 1<<4) {
    } else if (key < 1<<8) {
        u_int32_t low4Bits = key & 0x0000000f;
        key >>= 4;
        low4Bits <<= 4;
        key |= low4Bits;
    } else if (key < 1<<16) {
        u_int32_t low8bits = key & 0x000000ff;
        key >>= 8;
        low8bits <<= 8;
        key |= low8bits;
    } else {
        u_int32_t low16Bits = key & 0x0000ffff;
        key >>= 16;
        low16Bits <<= 16;
        key |= low16Bits;
    }
    printf("After hash key = %d.\n", key);
    return toBinary(key);
}

void ExtensiveHashTable::expand() {
    if (this->i >= 32) {
        printf("Buckets array length out of bounds.\n");
        return ;
    }
    Block** temp = new Block*[1<<((this->i)+1)]{NULL};
    if (temp == NULL) {
        printf(RED "allocate memory failed.\n" NONE);
        return ;
    }
    int k = 0, p = 0;
    (this->i)++;
    while (k < 1<<(this->i)) {
        int j = this->bucketArray[p]->j;
        for (int m = 0; m < (1<<(this->i-j)); m++) {
            temp[k+m] = this->bucketArray[p];
        }
        k += (1<<(this->i-j));
        p += (1<<(this->i-1-j));
    }
    delete[] (this->bucketArray);
    this->bucketArray = temp;
}

/*
 * When we need to split a block, this block may be connected by multiple
 * pointers, and we need to divide part of them to point to the new block.
 */
void ExtensiveHashTable::adjustNewBlock(Block* oldBlock, Block* newBlock, int location) {
    int count = this->i - oldBlock->j + 1;
    int k = 0, up = 0, temp = location;
    while (count) {
        int low1bit = location & 0x00000001;
        if (low1bit) {
            up += (1<<k);
        }
        k++;
        location >>= 1;
        count--;
    }
    int start = temp - up + (1<<(this->i-oldBlock->j));
    if (start < 0) {
        printf(RED "Adjust new block error\n" NONE);
        return ;
    }
    for (k = 0; k < (1<<(this->i-oldBlock->j)); k++) {
        printf(MAGENTA "Bucket Array[%d] = newBlock\n" NONE, start+k);
        this->bucketArray[start+k] = newBlock;
    }
}

ExtensiveHashTable::ExtensiveHashTable(u_int32_t size = 1) {
    this->i = size;
    bucketArray = new Block*[1<<i];
    for (int k = 0; k < 1<<i; k++) {
        bucketArray[k] = new Block();
    }
    printf("Initial an ExtensiveHashTable instance.\n");
}

int ExtensiveHashTable::findLocation(const u_int32_t key) {
    char* hashBin = hash(key);
    if (hashBin == NULL) {
        printf(RED "Hash error.\n" NONE);
        return -1;
    }
    int hashVal = toDecimal(hashBin, this->i);
    delete[] hashBin;
    if (hashVal >= (1<<(this->i)) || hashVal < 0) {
        printf(RED "Hash value error.\n" NONE);
        return -1;
    }
    return hashVal;
}

/* If no such record, return NULL.*/
Block* ExtensiveHashTable::findBlock(const u_int32_t key) {
    int hashVal = findLocation(key);
    if (hashVal == -1) {
        printf(RED "Find location error.\n" NONE);
        return NULL;
    }
    Block* p = this->bucketArray[hashVal];
    if (p == NULL) {
        printf(RED "No such block with key = %d.\n" NONE, key);
        return NULL;
    }
    return p;
}

Record* ExtensiveHashTable::search(const u_int32_t key, Block* b = NULL) {
    if (b == NULL) {
        b = findBlock(key);
    }
    if (b == NULL) {
        return NULL;
    }
    for (Record* r:(b->records)) {
        if (r != NULL && r->key == key) {
            printf("Searched record with key = %d, str = %s.\n", key, r->str);
            return r;
        }
    }
    printf("No such record with key = %d.\n", key);
    return NULL;
}

void ExtensiveHashTable::insertRecord(Record* r) {
    if (r == NULL) {
        printf("Can't insert NULL record!\n");
        return ;
    }
    Block* b = findBlock(r->key);
    if (b == NULL) {
        printf(RED "Exists NULL pointer in the bucket array.\n" NONE);
        return ;
    }
    if (search(r->key, b) != NULL) {
        printf(RED "Key duplicated, insertion terminated.\n" NONE);
        return ;
    }
    for (int k = 0; k < 4; k++) {
        if (b->records[k] == NULL) {
            b->records[k] = r;
            printf("Successfully inserted a record with key = %d in block %d\n", r->key, findLocation(r->key));
            //printf("this->i = %d\n", this->i);
            displayTable();
            return ;
        }
    }
    /*
     * No space in the current block.
     * Two opetions
     * */
    int j = b->j;
    int i = this->i;
    if (j == i) {
        printf(RED "Need to expand the buckey array.\n" NONE);
        printf(RED "Current table layout.\n" NONE);
        displayTable();
        printf("\n" NONE);
        expand();
        printf(RED "After expansion layout.\n" NONE);
        displayTable();
        i = this->i;
    }
    int hashVal = findLocation(r->key);
    printf("Location of key=%d is %s.\n", r->key, toReverseBinary(hashVal, this->i));
    if (j < i) {
        j++;
        printf("j = %d\n", j);
        Block* newBlock = new Block();
        if (newBlock == NULL) {
            printf(RED "New Block allcation error.\n" NONE);
            return ;
        }
        newBlock->j = j;
        b->j = j;
        /* Adjust the new block.*/
        adjustNewBlock(b, newBlock, hashVal);
        printf(CYAN "After adjustion.\n" NONE);
        displayTable();
        int temp = 0;
        /* Now we need to distribute these records into two blocks.*/
        printf(CYAN "To distribute all pointers in the current block.\n" NONE);
        for (int k = 0; k < 4; k++) {
            Record* re = b->records[k];
            if (re == NULL) {
                continue;
            }
            char* bin = hash(re->key);
            char c = bin[j-1];
            delete[] bin;
            if (c == '0') {
                printf("%d stays the same.\n", re->key);
                continue;
            } else {
                newBlock->records[temp++] = re;
                b->records[k] = NULL;
                printf("%d moves to the new block.\n", re->key);
            }
        }
        /*means all of the pointers go into one of the blocks*/
        if (temp == 0 || temp == 4) {
            printf("All records go into one of the blocks.\n");
            insertRecord(r);
        } else {
            printf("New block inserting.\n");
            char* bin = hash(r->key);
            char c = bin[j-1];
            if (c == '0') {
                for (int k = 0; k < 4; k++) {
                    if (b->records[k] == NULL) {
                        b->records[k] = r;
                        printf("Successfully inserted a record with key = %d in block %s\n", r->key, toReverseBinary(hashVal, this->i));
                        printf(CYAN "After distribution and insertion.\n" NONE);
                        displayTable();
                        return ;
                    }
                }
                printf("Record insertion failed.\n");
                return ;
            } else {
                newBlock->records[temp] = r;
                printf("Successfully inserted a record with key = %d in block %s\n", r->key, toReverseBinary(hashVal, this->i));
                printf(CYAN "After distribution and insertion.\n" NONE);
                displayTable();
                return ;
            }
        }
    }
}

void ExtensiveHashTable::deleteRecord(const u_int32_t key) {
    Block* b = findBlock(key);
    if (b == NULL) {
        printf("No such record.\n");
        return ;
    }
    for (int k = 0; k < 4; k++) {
        Record* re = b->records[k];
        if (re != NULL && re->key == key) {
            delete[] (re->str);
            delete re;
            b->records[k] = NULL;
            printf("Successfully deleted record with key = %d\n", key);
            return ;
        }
    }
    printf("No such record.\n");
}

void ExtensiveHashTable::displayTable() {
    printf("this->i = %d in displayTable.\n", this->i);
    for (int k = 0; k < (1<<(this->i)); k++) {
        char* bin = toReverseBinary(k, this->i);
        printf(MAGENTA "%s     ", bin);
        //printf(MAGENTA "%d    ", k);
        delete[] bin;
        if (this->bucketArray[k] == NULL) {
            printf("all null\n");
            continue;
        }
        for (Record* r: (this->bucketArray[k])->records) {
            if (r == NULL) {
                printf("null ");
            } else {
                printf("%d ", r->key);
            }
        }
        printf("j = %d\n" NONE, (this->bucketArray[k])->j);
    }
}

void test1() {
    ExtensiveHashTable eht;
    const char* s = "Hello World";
    /* Produce random numbers between 0 and 100 to test ExtensiveHashTable for 1000 times.*/
    printf("Insert Test.\n");
    for (int i = 0; i < 1000; i++) {
        printf("\n%d run.\n", i);
        u_int32_t key = (u_int32_t)(rand() % 1000);
        Record* r = new Record(key, s);
        eht.insertRecord(r);
    }
    printf("\n");
    eht.displayTable();
    printf("\nSearch Test.\n");
    for (int i = 0; i < 1000; i++) {
        int key = rand() % 1000;
        eht.search(key);
    }
    printf("Delete Test.\n");
    for (int i = 0; i < 1000; i++) {
        int key = rand() % 1000;
        eht.deleteRecord(key);
    }
}

void test2() {
}

int main() {
    test1();
}

File: code/extensive_hash_table.rs
/**********************************************
  > File Name		: extensive_hash_table.rs
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Wed 24 Feb 2021 09:00:06 PM CST
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

use std::ptr::NonNull;

struct Record {
    key: u32,
    s: String
}

impl Record {
    pub fn new(key: u32, s: String) -> Self {
        Record {
            key,
            s,
        }
    }
}

struct Block {
    j: i32,
    records: [Option<NonNull<Record>>; 4]
}

impl Block {
    pub fn new() -> Self {
        Block {
            j: 1,
            records: [None, None, None, None]
        }
    }
}

struct ExtensiveHashTable {
    i: i32,
    bucket_array: Vec<Option<NonNull<Block>>>
}

impl ExtensiveHashTable {
    fn hash(&self, mut key: u32) -> u32 {
        if key < 1<<4 {

        } else if key < 1<<8 {
            let mut low_4_bits = key & 0x0000000f;
            key >>= 4;
            low_4_bits <<= 4;
            key |= low_4_bits;
        } else if key < 1<<16 {
            let mut low_8_bits = key & 0x000000ff;
            key >>= 8;
            low_8_bits <<= 8;
            key |= low_8_bits;
        } else {
            let mut low_16_bits = key & 0x0000ffff;
            key >>= 16;
            low_16_bits <<= 16;
            key |= low_16_bits;
        }
        key
    }

    fn expand(&mut self) {
        if self.i >= 32 {
            println!("Buckets array length out of bounds.");
            return ;
        }
        let mut temp: Vec<Option<NonNull<Block>>> = Vec::with_capacity(1<<(self.i+1));
        temp.resize(1<<(self.i+1), None);
        let mut k = 0;
        let mut p = 0;
        self.i += 1;
        while k < 1<<self.i {
            let j:i32 = unsafe {
                (*self.bucket_array[p].unwrap().as_ptr()).j
            };
            for m in 0..(1<<(self.i-j)) {
                temp[k+m] = self.bucket_array[p].clone();
            }
            k += 1<<self.i-j;
            p += 1<<self.i-j-1;
        }
        self.bucket_array = temp;
    }

    /*
     * When we need to split a block, the block may be connected by multiple
     * pointers, and we need to have part of them point to a new block.
     */
    fn adjust_new_block(&mut self, old_block: Option<NonNull<Block>>, new_block: Option<NonNull<Block>>, mut location: i32) {
        let j = unsafe {
            (*old_block.unwrap().as_ptr()).j
        };
        let mut count = self.i + j + 1;
        let mut k = 0;
        let mut up = 0;
        let temp = location;
        while count > 0 {
            let low_1_bit = location & 0x00000001;
            if low_1_bit != 0 {
                up += 1<<k;
            }
            k += 1;
            location >>= 1;
            count -= 1;
        }
        let start = temp - up + 1<<self.i-j;
        if start < 0 {
            println!("Adjust new block error.");
            assert!(false);
        }
        for k in 0..(1<<self.i-j) {
            self.bucket_array[(start + k) as usize] = new_block.clone();
        }
    }

    fn find_location(&self, key: u32) -> u32 {
        let hashed_key = self.hash(key);
        let hash_val = hashed_key >> (32-self.i);
        if hash_val >= (1<<self.i) {
            println!("Hash value error.");
            assert!(false);
        }
        hash_val
    }

    fn find_block(&self, key: u32) -> Option<NonNull<Block>> {
        let hash_val = self.find_location(key);
        let p = self.bucket_array[hash_val as usize].clone();
        match &p {
            None => {
                println!("No such record with key = {}", key);
                None
            },
            Some(_) => {
                p
            }
        }
    }

    pub fn new() -> Self {
        ExtensiveHashTable {
            i: 1,
            bucket_array: {
                let mut b = Vec::with_capacity(2);
                b.resize(2, Some(unsafe {NonNull::new_unchecked(Box::into_raw(Box::new(Block::new())))}));
                b
            }
        }
    }

    fn search(&self, key: u32, mut b: Option<NonNull<Block>>) -> Option<NonNull<Record>> {
        if let None = b {
            b = self.find_block(key);
        }
        if let None = b {
            return None;
        }
        let records = unsafe {
            (*b.unwrap().as_ptr()).records
        };
        for record in records.iter() {
            if let Some(r) = record {
                if key == unsafe {(*r.as_ptr()).key} {
                    println!("Found record with key = {}", key);
                    return *record;
                }
            }
        }
        None
    }

    pub fn insert_record(&mut self, r: Option<NonNull<Record>>) {
        if let None = r {
            println!("Can't insert a None record!");
            return;
        }
        let key = unsafe {
            (*r.unwrap().as_ptr()).key
        };
        let b = self.find_block(key);
        if let None = b {
            panic!("Exists NULL pointer in the bucket array");
        }
        if let Some(_) = self.search(key, b) {
            println!("Key duplicated, insertion terminated");
            return ;
        }
        let records = &mut unsafe {
            (*b.unwrap().as_ptr()).records
        };
        for k in 0..4 {
            if let None = records[k] {
                records[k] = r;
                println!("Successfully inserted a record with key = {} in block {}", key, self.find_location(key));
                return ;
            }
        }
        /*
         * No space in the current block.
         */
        let mut j = unsafe {
            (*b.unwrap().as_ptr()).j
        };
        if j == self.i {
            println!("Need to expand the bucket array.");
            self.expand();
        }
        let hash_val = self.find_location(key);
        if (j < self.i) {
            j += 1;
            let mut new_block = Some(unsafe {NonNull::new_unchecked(Box::into_raw(Box::new(Block::new())))});
            unsafe {
                (*new_block.unwrap().as_ptr()).j = j;
                (*b.unwrap().as_ptr()).j = j;
            }
            self.adjust_new_block(b, new_block, hash_val as i32);
            /* 
             * Now we need to distribute the attributes in the old block into 
             * two blocks.
             */
            let mut temp = 0;//temp count for records moved to the new block.
            let new_block_records = &mut unsafe {
                (*new_block.unwrap().as_ptr()).records
            };
            for k in 0..4 {
                if let None = records[k] {
                    continue;
                }
                /*
                 * number j in the binary sequences decides which block to go
                 */
                let re_key = unsafe {
                    (*records[k].unwrap().as_ptr()).key
                };
                let hashed_key = self.hash(re_key);
                let d = hashed_key >> (j-1);
                if d & 0x1 == 0 {
                    println!("{} stays the same.", re_key);
                    continue;
                } else {
                    new_block_records[temp] = records[k].clone();
                    records[k] = None;
                    println!("{} moved to the new block.", re_key);
                }
            }
            /* If all of the pointers go to one of the blocks.*/
            if temp == 0 || temp == 4 {
                println!("All records go to one of the blocks.");
                self.insert_record(r);
            } else {
                let hashed_key = self.hash(key);
                let d = hashed_key >> (j-1);
                if d & 0x1 == 0 {
                    for k in 0..4 {
                        if let None = records[k] {
                            records[k] = r.clone();
                            println!("Successfully inserted a record with key = {} in block {}", key, self.find_location(key));
                            return ;
                        }
                    }
                } else {
                    new_block_records[temp] = r.clone();
                    println!("Successfully inserted a record with key = {} in block {}", key, self.find_location(key));
                }
            }
        }
    }
}

fn main() {
    
}
